File: 0091-decode-ways/0091-decode-ways.cpp
#include <string>
using namespace std;

class Solution {
public:
    int numDecodings(string s) {
        int n = (int)s.size();
        if (n == 0) return 0;

        // dp[0] = 1 (empty prefix), dp[1] depends on s[0]
        int prev2 = 1;
        int prev1 = (s[0] != '0') ? 1 : 0;

        for (int i = 2; i <= n; i++) {
            int cur = 0;

            // 1-digit decode: s[i-1] must not be '0'
            if (s[i - 1] != '0') {
                cur += prev1;
            }

            // 2-digit decode: s[i-2..i-1] must be between 10 and 26
            int two = (s[i - 2] - '0') * 10 + (s[i - 1] - '0');
            if (two >= 10 && two <= 26) {
                cur += prev2;
            }

            prev2 = prev1;
            prev1 = cur;
        }

        return prev1;
    }
};

File: 0091-decode-ways/0091-decode-ways.py
class Solution:
    def numDecodings(self, s: str) -> int:
        n = len(s)
        # dp[i] = number of ways to decode s[0:i]
        dp = [0] * (n + 1)
        dp[0] = 1  # empty string has exactly one way to decode (do nothing)
        dp[1] = 1 if s[0] != '0' else 0  # first character can't be '0' alone

        for i in range(2, n + 1):
            one_digit = s[i - 1]      # the single character just before position i
            two_digit = s[i - 2:i]    # the two characters just before position i

            if one_digit != '0':
                dp[i] += dp[i - 1]

            if '10' <= two_digit <= '26':
                dp[i] += dp[i - 2]

        return dp[n]    